### DISCUSSION TITLE
Configurable file backend for image cache

### DISCUSSION THREAD

**Discussion by artoale** [vercel/next.js#23006] [Ideas]
### Describe the feature you'd like to request

Currently, the image optimiser caches images locally, in the servers filesystem, typically under `.next/cache/images`. While this is fine on some platform, it's not uncommon for managed deployment target to restrict write access to the filesystem (e.g. Google App Engine only provides write access to an in-memory `/tmp` folder)

Giving users the option to write their own file backend would have a few benefits:
* less lock-in - people would be able to use the full power of next.js through more PaaS providers
* x-instances shared cache: in a multi-instances environment, the image cache could be shared by multiple next.js instances, allowing for even better performances and reduced disk usage
* industry standard: it's common practice to use third-party systems to handle image storage, for example AWS S3, or GCP Cloud Storage)

### Describe the solution you'd like

I'd like a hook that would allow me to provide a "filesystem backend".
The current implementation uses the following fs methods:
* `promises.mkdir`
* `promises.writeFile`
* `promises.readdir`
* `promises.unlink`
* `createReadStream`
* as well as the custom `fileExists` lib function

An ideal solution would allow users to provide custom implementation for those functions and default to use the node fs APIs otherwise

### Describe alternatives you've considered

Currently, the only option we have to use next image optimization within Google App Engine is to use a custom server, override the handler for `_next/images` with an handler that does exactly the same as the image optimizer (https://github.com/vercel/next.js/blob/99a4ea6e9a3a4a591d2a57c7b8f906d5ff30fd63/packages/next/next-server/server/image-optimizer.ts) but replacing fs calls with calls to a third-party provider (in our case Google Cloud Storage). This is not ideal, since it duplicates code, introduce opportunity for regression and would require manual changes when updating to newer version of Next.js

**Comment by artoale** [comment]
For further reference, here's an example from the python world for how a similar result is achieved by Django: https://docs.djangoproject.com/en/3.1/howto/custom-file-storage/

**Comment by Xetera** [comment]
I've been wondering about this one for a while too. I already use an s3 bucket for serving images so having to store cached images on disk re-introduces a lot of problems I solved by moving to an object storage. I believe extra egress costs on the server could also be mitigated if nextjs used a redirect to serve cached images directly through s3, although I'm not sure how well that would play with CORS policies.

I also have extra caching rules on the CDN hooked up to my s3 store which at best is annoying to have to replicate for nextjs images or not possible in the worst case scenario.
